### PR TITLE
Search and information about files

### PR DESCRIPTION
**Pull request by damian1996**
Adding classes File and DirectoryTree with basic features, such as get_info() for File, search files recursively from certain root node up to specified depth.

### CODE REVIEW

**Comment by graszka22** [comment]
- don't forget about file links - avoid looping infinitely
- don't forget about file permissions - you can get an error when you don't have permission to view the contents of the folder

**Comment by damian1996** [comment]
Check os.pardir instead of looking for a parent directory by hand.
And check if dates for directory are OK (last modification, last opened).

File: api.txt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+
+Partition(*args, **kwargs) - uneccessary for you
+
+get_name() - returns name of partition/device
+
+get_mountpoint() - returns mountpoint for partition
+
+get_total() - returns total for partition
```
Comment (graszka22) [inline api.txt]: total… of what? size?

Comment (graszka22) [reply api.txt]: it's not in the `Partition.py` o.o?

File: src/Disks.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        partitions = self.find_partitions()
+        partitions_usage = [ps.disk_usage(prt.mountpoint) for prt in partitions]
+        return list(zip(partitions, partitions_usage))
+
+    def create_partitions_info(self):
+        info = self.disk_usage_for_partitions()
+        return [Partition(partition, usage) for partition, usage in info]
+
+    def get_partitions(self):
+        for prt in self.partitions:
```
Comment (graszka22) [inline src/Disks.py]: returning an iterator here doesn't make any sense, since `self.partitions` is already calculated list 
returning an iterator makes sense when we're iterating over a large dataset and we don't want to store all of it in the memory or when it's lazily calculated
a good use case for an iterator would be, as I was mentioning before, walking over a file tree (it's big data set and it would be nice if it was calculated lazily - yeah, I know, it's impossible because of system api) but here you have a list with a few elements and it's just unecessary complication

File: src/Disks.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    
+    def find_partitions(self):
+        partitions = ps.disk_partitions()
+        partitions = [prt for prt in partitions if not prt.mountpoint.startswith('/snap')]
+        return partitions
+
+    def disk_usage_for_partitions(self):
+        partitions = self.find_partitions()
+        partitions_usage = [ps.disk_usage(prt.mountpoint) for prt in partitions]
+        return list(zip(partitions, partitions_usage))
```
Comment (graszka22) [inline src/Disks.py]: you can write it like this:
```python
return [(prt, ps.disk_usage(prt.mountpoint)) for prt in self.find_partitions()]
```

File: src/Partition.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    def __init__(self, partition_info, partition_usage):
+        self.device = partition_info.device
+        self.mountpoint = partition_info.mountpoint
+        self.total = partition_usage.total
+        self.used = partition_usage.used
+        self.free = partition_usage.free
+        self.percents_usage = partition_usage.percent
+    
+    def __repr__(self):
+        return 'Device {}, mounted at {}. Used memory is {}% of {} bytes' \
```
Comment (graszka22) [inline src/Partition.py:13]: remember f-strings? :heart:

File: src/File.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import os, time
+
+class File:
+    def __init__(self, name, path):
+        self.name = name
+        self.path = path
+        self.file_type = os.path.isfile(path)
```
Comment (graszka22) [inline src/File.py]: a name `file_type` should be a… file type, not a boolean value -,- a better naming would be simply `is_file` :stuck_out_tongue:

File: src/File.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import os, time
+
+class File:
+    def __init__(self, name, path):
```
Comment (graszka22) [inline src/File.py]: does `path` contain a name? it's not clear… it's better to give just a `path` to the constructor and let the constructor figure out the filename!

File: src/File.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        self.path = path
+        self.file_type = os.path.isfile(path)
+        self.size, self.parent_size = 0, 0
+        self.next_files = []
+
+    def __repr__(self):
+        return "File {}".format(self.name)
+
+    def convert_epoch_time_to_date(self, epoch_time):
+        return time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(epoch_time))
```
Comment (graszka22) [inline src/File.py]: you can put this in separate file named `utils.py` or something like that, this function is not related to `File` class

Comment (graszka22) [reply src/File.py]: also the `File` class should give me the time as a number, not converted to a string - I can convert it myself and conversion to string can cause me a problem when I want to sort the dates etc.
and when I want to display it I'd use an utility function
also, if we'd want to translate the app (highly unlikely xd I'm just writing this so you'd know for the future) there are many date formats so displaying a date is part of internationalization and is done at the and, when displaying the date

File: src/File.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    
+    def get_file_info(self):
+        count_files, count_subdirs = self.count_files_and_subdirs()
+        return {
+            "name": self.get_name(),
+            "path": self.get_path(),
+            "type": self.get_file_type(),
+            "size": self.size,
+            "when_created": self.get_creation_time(),
```
Comment (graszka22) [inline src/File.py]: "creation_date" sounds better :stuck_out_tongue:

File: src/File.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            "last_modified": self.get_last_modified_time(),
+            "last_access": self.get_last_opened_time(),
+            "count_files": count_files,
+            "count_subdirectories": count_subdirs,
+            "size_parent_directory": self.parent_size,
```
Comment (graszka22) [inline src/File.py]: "parent_directory_size"?

File: src/File.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        return {
+            "name": self.get_name(),
+            "path": self.get_path(),
+            "type": self.get_file_type(),
+            "size": self.size,
+            "when_created": self.get_creation_time(),
+            "last_modified": self.get_last_modified_time(),
+            "last_access": self.get_last_opened_time(),
+            "count_files": count_files,
+            "count_subdirectories": count_subdirs,
```
Comment (graszka22) [inline src/File.py]: "subdirectiories_count"?

File: src/File.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        count_files, count_subdirs = self.count_files_and_subdirs()
+        return {
+            "name": self.get_name(),
+            "path": self.get_path(),
+            "type": self.get_file_type(),
+            "size": self.size,
+            "when_created": self.get_creation_time(),
+            "last_modified": self.get_last_modified_time(),
+            "last_access": self.get_last_opened_time(),
+            "count_files": count_files,
```
Comment (graszka22) [inline src/File.py]: "files_count"?

File: src/File.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        return self.file_type
+    
+    def get_size(self):
+        return self.size
+
+    def get_size_parent_directory(self):
+        return self.parent_size
+
+    def count_files_and_subdirs(self):
+        if os.path.isfile(self.path):
```
Comment (graszka22) [inline src/File.py]: you know it already… it's in `file_type`…

File: src/File.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            return 0, 0
+        all_files_in_dir = len(os.listdir(self.path))
+        onlyfiles = sum(os.path.isfile(os.path.join(self.path, f)) for f in os.listdir(self.path))
```
Comment (graszka22) [inline src/File.py]: why calling `os.listdir` twice? also you have `self.next_files`… maybe you can use that and don't do any syscalls here at all?

File: api.txt (hunk 1 above)
Comment (damian1996) [reply api.txt]: > total… of what? size?

yep, size

> it's not in the `Partition.py` o.o?

I've had this method in Partition.py under other name. Fixed.

File: src/Disks.py (hunk 2 above)
Comment (damian1996) [reply src/Disks.py]: Ok, I see. I've changed it. Now method get_partitions returns a list of partitions.

File: src/Disks.py (hunk 3 above)
Comment (damian1996) [reply src/Disks.py]: Ok, I've changed it.

Comment (damian1996) [reply src/Disks.py]: And I've made one-liners in another two methods in class Disks.py

File: src/Partition.py (hunk 4 above)
Comment (damian1996) [reply src/Partition.py:13]: I didn't forget about f-strings, but I have too old python version, which doesn't support f-strings, so I've used format.

File: src/File.py (hunk 5 above)
Comment (damian1996) [reply src/File.py]: Corrected.

File: src/File.py (hunk 12 above)
Comment (damian1996) [reply src/File.py]: OK

File: src/File.py (hunk 11 above)
Comment (damian1996) [reply src/File.py]: OK

File: src/File.py (hunk 10 above)
Comment (damian1996) [reply src/File.py]: OK

File: src/File.py (hunk 9 above)
Comment (damian1996) [reply src/File.py]: OK

File: src/File.py (hunk 7 above)
Comment (damian1996) [reply src/File.py]: Ok, I will remove convert_epoch_time_to_date and I will return epoch_time (as integer) and you will be responsible for conversion in the future.

File: src/File.py (hunk 8 above)
Comment (damian1996) [reply src/File.py]: OK

File: src/File.py (hunk 13 above)
Comment (damian1996) [reply src/File.py]: Done

File: src/File.py (hunk 6 above)
Comment (damian1996) [reply src/File.py]: Right, done.

File: api.txt [hunk 14]
```diff
@@ -0,0 +1,53 @@
+Disks() - constructor
+
+get_partitions() - generator yielding following get_partitions
```
Comment (graszka22) [inline api.txt]: the `api.txt` needs to be updated
`get_partitions()` is no longer a generator, `File` has other constructor, etc :p

File: src/DirectoryTree.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+                child.set_parent_size(dir_size)
+            files_dict[dirname] = root
+        
+        files_dict[path].set_parent_size(0)
+        return files_dict[path], files_dict
+
+    def check_access(self, path):
+        return os.access(path, os.R_OK)
+
+    def file_to_skip(self, path):
```
Comment (graszka22) [inline src/DirectoryTree.py]: better name would be something like `should_skip_file`, `file_to_skip` suggests that the function returns a file to skip o.o?

File: src/DirectoryTree.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        return not self.check_access(path) or os.path.islink(path)
+
+    def get_parent(self, path):
+        return os.path.dirname(path) if os.path.isfile(path) else os.path.abspath(os.path.join(path, ".."))
```
Comment (graszka22) [inline src/DirectoryTree.py]: `os.path.dirname` works also when `path` is not a file, no need to check - it works simply by dropping the last part of the path

File: src/DirectoryTree.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+    def walklevel(self, path=''):
+        path = path if path != '' else self.path
+        if path.startswith(("~", "/")):
+            path = os.path.expanduser(path)
+
+        root = self.files[path]
+        for child in root.get_children():
```
Comment (graszka22) [inline src/DirectoryTree.py]: no need for generator here, you already have the list

File: src/DirectoryTree.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+                raise Exception('You have not access to {} or directory does not exists'.format(dir_path))
+            dir_to_remove = self.files[dir_path]
+            size_of_removed_directory = dir_to_remove.get_size()
+            keys = self.files.keys()
+            path_copy = str(dir_path)
+            while path_copy != '/':
+                path_copy = os.path.abspath(os.path.join(path_copy, '..'))
+                if not path_copy in keys:
+                    break
+                print(path_copy)
```
Comment (graszka22) [inline src/DirectoryTree.py]: remove print

File: src/DirectoryTree.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+            path = os.path.expanduser(path)
+
+        root = self.files[path]
+        for child in root.get_children():
+            yield child
+
+    def remove_file(self, path):
+        try:
+            if not self.check_access(path):
+                raise Exception('You have not access to {} or file does not exists'.format(path))
```
Comment (graszka22) [inline src/DirectoryTree.py]: "you don't have access"

Comment (graszka22) [reply src/DirectoryTree.py]: "or file does not exist~s~"

File: src/DirectoryTree.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+                        parent_dir.delete_child(i)
+            os.remove(path)
+        except KeyError:
+            raise KeyError('File {} does not exists'.format(path.split('/')[-1]))
+        except Exception as e:
+            raise Exception(e)
+        
+    def remove_directory(self, dir_path):
+        try:
+            if not self.check_access(dir_path):
```
Comment (graszka22) [inline src/DirectoryTree.py]: you've repeated 20 lines of code, try to extract it to another function
you can name it `remove_tree_node` or sth like that because I can see that all it does is traversing the tree upwards and correcting the sizes

File: src/DirectoryTree.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+        return os.access(path, os.R_OK)
+
+    def file_to_skip(self, path):
+        return not self.check_access(path) or os.path.islink(path)
+
+    def get_parent(self, path):
+        return os.path.dirname(path) if os.path.isfile(path) else os.path.abspath(os.path.join(path, ".."))
+
+    def walklevel(self, path=''):
+        path = path if path != '' else self.path
```
Comment (graszka22) [inline src/DirectoryTree.py]: `path = path or self.path`?

File: src/Main.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
+from DirectoryTree import DirectoryTree
+import unittest, os
+from File import File
+from Disks import Disks
+
+def get_files_by_level2(gen):
+    while True:
```
Comment (graszka22) [inline src/Main.py]: just FYI you can iterate like this:
```python
for file in gen:
    print(file.get_file_info())
```

File: src/File.py [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+    def __repr__(self):
+        return "File {}".format(self.name)
+
+    def get_file_info(self):
+        files_count, count_subdirs = self.count_files_and_subdirs()
+        return {
+            "name": self.get_name(),
+            "path": self.get_path(),
+            "type": self.if_is_file(),
```
Comment (graszka22) [inline src/File.py]: I've said it previously that "type" is not appropriate name :p

File: src/DirectoryTree.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+            size_of_removed_file = file_to_remove.get_size()
+            keys = self.files.keys()
+            path_copy = str(path)
+            while path_copy != '/':
+                path_copy = os.path.abspath(os.path.join(path_copy, '..'))
+                if not path_copy in keys:
+                    break
+                self.files[path_copy].decrease_sizes(size_of_removed_file)
+            name = file_to_remove.get_name()
+            parent_path = os.path.abspath(os.path.join(path, '..'))
```
Comment (graszka22) [inline src/DirectoryTree.py]: funny trick :smile:  :+1:

File: src/DirectoryTree.py (hunk 15 above)
Comment (damian1996) [reply src/DirectoryTree.py]: I will change it to is_file_to_skip.

File: src/DirectoryTree.py (hunk 17 above)
Comment (damian1996) [reply src/DirectoryTree.py]: Done

File: src/DirectoryTree.py (hunk 18 above)
Comment (damian1996) [reply src/DirectoryTree.py]: Done

File: src/DirectoryTree.py (hunk 20 above)
Comment (damian1996) [reply src/DirectoryTree.py]: Done. Additionally, I divided this one method to smaller methods.

File: src/Main.py (hunk 22 above)
Comment (damian1996) [reply src/Main.py]: Yep, it will be simpler.

File: src/DirectoryTree.py (hunk 24 above)
Comment (damian1996) [reply src/DirectoryTree.py]: :+1:

File: src/File.py (hunk 23 above)
Comment (damian1996) [reply src/File.py]: Okay, I will change it to "is_file".

File: src/DirectoryTree.py (hunk 19 above)
Comment (damian1996) [reply src/DirectoryTree.py]: :+1:

File: src/DirectoryTree.py (hunk 21 above)
Comment (damian1996) [reply src/DirectoryTree.py]: :+1:

File: src/DirectoryTree.py (hunk 16 above)
Comment (damian1996) [reply src/DirectoryTree.py]: Oh, indeed... :+1:

**Review by graszka22 — APPROVED** [review]
